src/lib.rs

```diff
@@ -73,46 +49,447 @@ Note that `--exclude` has to be specified in conjunction with the `--all` flag.
+				$(
+					if let Some(value) = &$option {
+						values.insert($name, vec![value.into()]);
+					}
+				)*
+			};
+		}
+
+		macro_rules! add_slices {
+			($($name:expr => $slice:expr),* $(,)?) => {
+				$(
+					if !$slice.is_empty() {
+						values.insert($name, $slice.iter().map(Into::into).collect());
+					}
+				)*
+			};
+		}
+
+		add_bools! {
+			 "all"                 => self.all,
+			 "lib"                 => self.lib,
+			 "bins"                => self.bins,
+			 "examples"            => self.examples,
+			 "tests"               => self.tests,
+			 "benches"             => self.benches,
+			 "all-targets"         => self.all_targets,
+			 "release"             => self.release,
+			 "all-features"        => self.all_features,
+			 "no-default-features" => self.no_default_features,
+		}
+
+		add_options! {
+			"jobs"       => self.jobs,
+			"profile"    => self.profile,
+			"target"     => self.target,
+		}
+
+		add_slices! {
+			"package"  => self.package,
+			"exclude"  => self.exclude,
+			"bin"      => self.bin,
+			"example"  => self.example,
+			"test"     => self.test,
+			"bench"    => self.bench,
+			"features" => self.features,
+		}
+
+		fake_arg_matches(values).compile_options(config, mode, workspace)
+    }
 }
 
-impl std::fmt::Debug for StrErr {
-	fn fmt(&self, f :&mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
-		// Difference of this debug impl to the one provided by the derive macro
-		// is that special chars like newlines and " aren't escaped.
-		// We have some human-readable errors where newlines help with the output.
-		write!(f, "StrErr(\"{}\")", self.0)
+fn fake_arg_matches(values :HashMap<&'static str, Vec<OsString>>) -> impl ArgMatchesExt {
+	struct FakeArgMatches(HashMap<&'static str, Vec<OsString>>);
+
+	impl ArgMatchesExt for FakeArgMatches {
+		fn _value_of(&self, name :&str) -> Option<&str> {
+			let value = self
+				.0
+				.get(name)?
+				.get(0)?
+				.to_str()
+				.expect("unexpected invalid UTF-8 code point");
+			Some(value)
+		}
+
+		fn _value_of_os(&self, name :&str) -> Option<&OsStr> {
+			self.0.get(name)?.get(0).map(Deref::deref)
+		}
+
+		fn _values_of(&self, name :&str) -> Vec<String> {
+			self.0
+				.get(name)
+				.map(Deref::deref)
+				.unwrap_or_default()
+				.iter()
+				.map(|value| {
+					value
+						.to_str()
+						.expect("unexpected invalid UTF-8 code point")
+						.to_owned()
+				})
+				.collect()
+		}
+
+		fn _values_of_os(&self, name :&str) -> Vec<OsString> {
+			self.0
+				.get(name)
+				.map(Deref::deref)
+				.unwrap_or_default()
+				.iter()
+				.map(Clone::clone)
+				.collect()
+		}
+
+		fn _is_present(&self, name :&str) -> bool {
+			self.0.contains_key(name)
+		}
 	}
+
+	FakeArgMatches(values)
 }
 
 struct ExecData {
-	cargo_exe :Option<String>,
 	supports_color :bool,
 	relevant_cmd_infos :Vec<CmdInfo>,
 }
 
 impl ExecData {
 	fn new(supports_color :bool) -> Self {
-		let cargo_exe = which::which("cargo")
```

Comment:
Why have you removed this?

Reply:
I remove the parts while struggling to rewrite `testsuite` and committed inadvertently. Without this, `$CARGO` is set to `current_exe()` (i.e. `~/.cargo/bin/cargo-udeps`).

But shouldn't this `std::env::var(cargo::CARGO_ENV)`? `which::which("cargo")` points at wrong version of `cargo` unless `nightly` is the default toolchain. (`$(which cargo)`→ `~/.cargo/bin/cargo` → `$(rustup which cargo)`) `$CARGO` should be always present since `cargo-udeps` is a cargo subcommand. And we can try it with `cargo +nightly run -- udeps --manifest-path /path/to/crate/Cargo.toml`.

Reply:
> Without this, $CARGO is set to current_exe() (i.e. ~/.cargo/bin/cargo-udeps).

That's a problem for build scripts which read `$CARGO` and then do `$CARGO build`. cargo-udeps doesn't support build subcommand :).